### PR TITLE
Op 3 create ecs deployment method datastore controller

### PR DESCRIPTION
**Pull request by gchuev-opsfleet**
- Added CloudFormation templates aws-ecs deployments.
- Updated Terraform module for aws-ecs deployments.

### CODE REVIEW

**Comment by idanrookout** [comment]
@alexeygutkin please review, talk to @gchuev-opsfleet if needed

File: aws-ecs/terraform/main.tf [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    certificate_arn           = null
+    task_cpu                  = 512
+    task_memory               = 1024
+    container_cpu             = 256
+    container_memory          = 512
+    storage_size              = 20
+    container_port            = 8080
+    load_balancer_port        = 443
+  }
+  controller_settings = merge(local.default_controller_settings, var.controller_settings)
```
Comment (alexeygutkin) [inline aws-ecs/terraform/main.tf:41]: the default values already declared in variables.tf. why declare it twice and merge?

File: aws-ecs/terraform/variables.tf [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+variable "launch_type" {
+  description = "Launch type for ECS Services. Set to EC2 if you've provided cluster_name variable with existing EC2 cluster name."
+  type        = string
+  default     = "FARGATE"
+}
+
+variable "controller_settings" {
+  type = map(string)
```
Comment (alexeygutkin) [inline aws-ecs/terraform/variables.tf:43]: may be reasonable to use  `structural type` here https://www.terraform.io/language/expressions/type-constraints#structural-types

File: aws-ecs/terraform/variables.tf [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    task_cpu                  = 512
+    task_memory               = 1024
+    container_cpu             = 256
+    container_memory          = 512
+    container_port            = 7488
+    load_balancer_port        = 7488
+  }
+}
+
+variable "datastore_settings" {
```
Comment (alexeygutkin) [inline aws-ecs/terraform/variables.tf:63]: may be reasonable to use  `structural type` here https://www.terraform.io/language/expressions/type-constraints#structural-types

File: aws-ecs/terraform/templates/controller_tls_task_def.tpl [hunk 4]
```diff
@@ -0,0 +1,87 @@
+[
+  {
+    "name": "s3-downloader",
+    "image": "amazon/aws-cli:latest",
+    "cpu": 128,
+    "memory": 128,
+    "essential": false,
+    "command": ["s3","sync", "s3://${certificate_bucket_name}/${certificate_bucket_prefix}", "/var/controller-tls-secrets"],
```
Comment (alexeygutkin) [inline aws-ecs/terraform/templates/controller_tls_task_def.tpl:8]: why s3 for certificates? why not ACM?

**Review by alexeygutkin — CHANGES_REQUESTED** [review]

File: aws-ecs/terraform/variables.tf (hunk 3 above)
Comment (gchuev-opsfleet) [reply aws-ecs/terraform/variables.tf:63]: Same thing as with default map merge. You can check last post here:
https://discuss.hashicorp.com/t/map-variable-default-value-behavior-change/1646/3

File: aws-ecs/terraform/templates/controller_tls_task_def.tpl (hunk 4 above)
Comment (gchuev-opsfleet) [reply aws-ecs/terraform/templates/controller_tls_task_def.tpl:8]: It's build-in(container level) certificate for TLS server-mode, so there should be a file, we cannot export ACM certificates as file and it's optional.
Other option could be SSM parameter/secrets manager for example. I just thought that as we're using files in Docker/Helm examples so let it be here as well.
For ACM certificate used at load balancer level there's another  variable option used in datastore/controller_settings map.

File: aws-ecs/terraform/main.tf [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    task_memory               = 1024
+    container_cpu             = 256
+    container_memory          = 512
+    container_port            = 7488
+    load_balancer_port        = 7488
+  }
+  default_datastore_settings = {
+    deploy                    = true
+    container_name            = "rookout-datastore"
+    serer_mode                = "PLAIN"
```
[the remark this answers is not in the document]
Comment (gchuev-opsfleet) [reply aws-ecs/terraform/main.tf]: fixed

File: aws-ecs/cloudformation/existing-vpc.dev.json [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        "ParameterKey": "VpcId",
+        "ParameterValue": "vpc-0ddd4385198044b6f"
+    },
+    {
+        "ParameterKey": "PrivateSubnets",
+        "ParameterValue": "subnet-02b76d71542167f00,subnet-0b24284913b02c8d8"
+    },
+    {
+        "ParameterKey": "PublicSubnets",
+        "ParameterValue": "subnet-02b76d71542167f00,subnet-0b24284913b02c8d8"
```
[the remark this answers is not in the document]
Comment (gchuev-opsfleet) [reply aws-ecs/cloudformation/existing-vpc.dev.json]: done

File: aws-ecs/cloudformation/existing-vpc.dev.json [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        "ParameterKey": "CreateLB",
+        "ParameterValue": "true"
+    },
+    {
+        "ParameterKey": "VpcId",
+        "ParameterValue": "vpc-0ddd4385198044b6f"
+    },
+    {
+        "ParameterKey": "PrivateSubnets",
+        "ParameterValue": "subnet-02b76d71542167f00,subnet-0b24284913b02c8d8"
```
[the remark this answers is not in the document]
Comment (gchuev-opsfleet) [reply aws-ecs/cloudformation/existing-vpc.dev.json]: done

File: aws-ecs/cloudformation/existing-vpc.dev.json [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        "ParameterKey": "Environment",
+        "ParameterValue": "dev"
+    },
+    {
+        "ParameterKey": "CreateLB",
+        "ParameterValue": "true"
+    },
+    {
+        "ParameterKey": "VpcId",
+        "ParameterValue": "vpc-0ddd4385198044b6f"
```
[the remark this answers is not in the document]
Comment (gchuev-opsfleet) [reply aws-ecs/cloudformation/existing-vpc.dev.json]: done

File: aws-ecs/cloudformation/existing-vpc.dev.json [hunk 9]
```diff
@@ -0,0 +1,58 @@
+[
+    {
+        "ParameterKey": "TokenSecretArn",
+        "ParameterValue": "arn:aws:secretsmanager:us-east-1:782824484157:secret:rookout_token-JeQYA2"
```
[the remark this answers is not in the document]
Comment (gchuev-opsfleet) [reply aws-ecs/cloudformation/existing-vpc.dev.json]: done

**Comment by gchuev-opsfleet** [comment]
@alexeygutkin pushed fixes, please check comments.

**Pull request closed without merging**